File: py/1502.can-make-arithmetic-progression-from-sequence.py
#
# @lc app=leetcode id=1502 lang=python3
#
# [1502] Can Make Arithmetic Progression From Sequence
#

# @lc code=start
class Solution:
    def canMakeArithmeticProgression(self, arr: List[int]) -> bool:
        max_val = min_val = arr[0]
        for n in arr:
            if n < min_val:
                min_val = n
            elif n > max_val:
                max_val = n
        width = len(arr)
        if (max_val-min_val) % (width-1):
            return False
        diff = (max_val-min_val)//(width-1)
        if diff == 0:
            return True
        for n in arr:
            if n > 1000000:
                n -= 3000000
            if (n-min_val) % diff:
                return False
            pos = (n-min_val)//diff
            if pos >= width or arr[pos] > 1000000:
                return False
            arr[pos] += 3000000
        return True


# @lc code=end

File: 1502.can-make-arithmetic-progression-from-sequence.cpp
/*
 * @lc app=leetcode id=1502 lang=cpp
 *
 * [1502] Can Make Arithmetic Progression From Sequence
 */
#include <vector>
#include <climits>
using namespace std;
// @lc code=start
// T:O(N) S:O(1)
class Solution
{
public:
    bool canMakeArithmeticProgression(vector<int> &arr)
    {
        int min_val = arr[0], max_val = arr[0];
        for (auto n : arr)
        {
            if (n < min_val)
            {
                min_val = n;
            }
            else if (n > max_val)
            {
                max_val = n;
            }
        }
        int width = arr.size();
        if ((max_val - min_val) % (width - 1))
        {
            return false;
        }
        int diff = (max_val - min_val) / (width - 1);
        if (diff == 0)
        {
            return true;
        }
        for (auto n : arr)
        {
            if (n > 1000000)
            {
                n -= 3000000;
            }
            if ((n - min_val) % diff)
            {
                return false;
            }
            int pos = (n - min_val) / diff;
            if (pos >= width || arr[pos] > 1000000)
            {
                return false;
            }
            arr[pos] += 3000000;
        }
        return true;
    }
};
// @lc code=end
